sdk/core/azure-core/inc/azure/core/azure_assert.hpp

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Provide assert macros to use as pre-conditions.
+ *
+ * @remark Asserts are turned off is `NDEBUG` is defined
+ *
+ */
+
+#pragma once
+
+// uncomment next line to disable asserts
+// #define NDEBUG
+#include <cassert>
+
+#define AZURE_ASSERT(exp) assert(exp)
+#define AZURE_ASSERT_MSG(exp, msg) assert(((void)msg, exp))
+#define AZURE_UNREACHABLE_CODE AZURE_ASSERT_MSG(false, "Unreachable code was access")
+#define AZURE_NOT_IMPLEMENTED AZURE_ASSERT_MSG(false, "Not Implemented")
```

Comment:
The problem is assert is not equal to abort.
assert is a no-op in non-debug build.

Reply:
It's like, we crash on non-implemented function in debug build, while we don't crash in release build.

Reply:
That's the proposal.  We want people to find the bugs on Debug and fix their code before going to production.
The asserts are only allowed for things that can be fixed before going to production.  We don't want to add Assertions to runtime errors, things coming from I/O, etc.

For example, anyone calling Value() from a Nullable<T> can fix their code by first checking HasValue(), 

Also, people can define what assert can do on Production build.  By default it is no-op, but when NDEBUG is set, people can define assert macro to do what they want.

Reply:
I like the assert and then abort().  That way we get a good debug experience and retail builds quickly terminate. abort() should happen for the cases where contracts are broken.  A contract by definition is somewhere that continuing on will lead to a crash.

Reply:
> That's the proposal. We want people to find the bugs on Debug and fix their code before going to production.
The asserts are only allowed for things that can be fixed before going to production. We don't want to add Assertions to runtime errors, things coming from I/O, etc.

The behavior on this kind of error is defined in our [Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html#cpp-errors). So you're proposing our guidelines doesn't apply on release build?

Reply:
> Also, people can define what assert can do on Production build. By default it is no-op, but when NDEBUG is set, people can define assert macro to do what they want.

Release/Debug build is not as distinct on Linux as in Windows Visual Studio. Sometimes people just want to write a small program, they just use `g++ some_source.cpp` for simplicity, or with `-g` flag for debug. `NDEBUG` macro isn't set in either case.

Reply:
I don't think we support people not using CMake project to consume us.  And sure, you can play around and create quicks apps for learning, but that's not likely someone who will create a real-world product and charge people for it.

Reply:
@RickWinter I have read many times your comment but I still don't understand what are you asking/suggesting?
Is it:
- Do not use assert for `not implemented` errors?   if yes, that's needs to be fixed out of this change, since that's currently happening (we are aborting on non-implemented methods as a way to tell people that's not something they want to call :)

Reply:
> I don't think we support people not using CMake project to consume us.

That's a big decision I didn't know before.

Let's say our customer installed storage sdk via vcpkg, and then he works on a project with Visual Studio, he directly creates a .sln solution without CMake. He's been using Microsoft stuff the whole time. Is this scenario supported by us?

Another scenario on Linux is: customers build our sdk from source by themselves and install headers and libs to system path. Now they can use whatever build tools they like. This is also not supported by us?

Reply:
> Let's say our customer installed storage sdk via vcpkg, and then he works on a project with Visual Studio, he directly creates a .sln solution without CMake. He's been using Microsoft stuff the whole time. Is this scenario supported by us?
>
> Another scenario on Linux is: customers build our sdk from source by themselves and install headers and libs to system path. Now they can use whatever build tools they like. This is also not supported by us?

End-users can do and try any options they want. That is different to what we support and demonstrate on our samples and tests.  I don't think we should provide documentation and validate all possible ways of consumption, instead, we provide one which is CMake as OS - agnostic.

If any client has a different setup, they can always create an issue to ask for support (We recently hear from someone building for iOS). Then we can decide if would help/guide or just mention that's not expected/supported (management should tell)